### PR TITLE
Permission

### PR DESCRIPTION
**Pull request by btkostner**
Adds permissions based on user's teams and organizations. Sane defaults based on elementary organization teams. Config option to bypass permissions for testing and hacking on.

### CODE REVIEW

**Comment by btkostner** [comment]
#95

File: app/models/user.js [hunk 1]
```diff
@@ -29,12 +29,47 @@ UserSchema.statics.updateOrCreate = function(accessToken, profile) {
         avatar:   profile._json['avatar_url'],
         github:   {accessToken: accessToken},
         joined:   Date.now(),
-        active:   true,
       });
     }
+  })
+  .then(getRights)
+  .then(user => {
+    console.log(user);
```
Comment (aroman) [inline app/models/user.js]: did you mean to take this out?

File: app/models/user.js [hunk 2]
```diff
@@ -29,12 +29,47 @@ UserSchema.statics.updateOrCreate = function(accessToken, profile) {
...
+    return user
   });
 };
 
+// Checks github organizations and teams for correct permissions
+function getRights(user) {
+  let gh = new Hubkit({ token: user.github.accessToken, boolean: true });
```
Comment (aroman) [inline app/models/user.js]: const

File: app/models/user.js (hunk 1 above)
Comment (btkostner) [inline app/models/user.js]: woops. yes.

File: app/models/user.js [hunk 3]
```diff
@@ -29,12 +29,47 @@ UserSchema.statics.updateOrCreate = function(accessToken, profile) {
...
+
+  return Promise.all([
+    gh.request(`GET /orgs/${CONFIG.RIGHTS_ORG}/members/${user.username}`),
+    gh.request(`GET /teams/${CONFIG.RIGHTS_REVIEWER}/memberships/${user.username}`),
+    gh.request(`GET /teams/${CONFIG.RIGHTS_ADMIN}/memberships/${user.username}`),
+  ])
+  .then((membership) => {
```
Comment (aroman) [inline app/models/user.js]: should use destructuring syntax here, like:

```js
.then(([isBeta, isReviewer, isAdmin]) => { ...
```

File: app/models/user.js [hunk 4]
```diff
@@ -29,12 +29,47 @@ UserSchema.statics.updateOrCreate = function(accessToken, profile) {
...
+      user.rights = 'beta'
+    } else {
+      user.rights = 'user'
+    }
+    return user.save();
+  });
+}
+
+function rightsValidator(value) {
+  var valid = ['user', 'beta', 'reviewer', 'admin'];
```
Comment (aroman) [inline app/models/user.js]: i'd probably replace this entire function with an inline closure, like:

```js
{validate: v => ['user', 'beta', 'reviewer', 'admin'].indexOf(v) >= 0}
```

File: config.example.json [hunk 5]
```diff
@@ -7,5 +7,9 @@
   "JENKINS_ENABLED": false,
   "JENKINS_URL": "http://<jenkins username>:<jenkins user api token>@jenkins.elementaryos.org",
   "JENKINS_JOB": "deb-new-test",
-  "JENKINS_SECRET": "test"
+  "JENKINS_SECRET": "test",
+  "RIGHTS_ENABLED": false,
+  "RIGHTS_ADMIN": 213128,
```
Comment (aroman) [inline config.example.json]: what are these numbers?

Comment (btkostner) [inline config.example.json]: github team ids

**Comment by aroman** [comment]
like_fedex

**Pull request merged**